### PR TITLE
Support RDB Short Read in RediSearch (a.k.a diskless-load)

### PR DESCRIPTION
**Pull request by oshadmi**

### CODE REVIEW

**Review by ashtul — DISMISSED** [review]
We can check `RedisModule_IsIOError` only at the end instead of after each `RedisModule_Load...` call. This will make the code more readable.
Can a macro help readability?

File: src/spec.c [hunk 1]
```diff
@@ -1743,6 +1755,9 @@ int Indexes_RdbLoad(RedisModuleIO *rdb, int encver, int when) {
     }
   }
   return REDISMODULE_OK;
+
+cleanup:
+  return REDISMODULE_ERR;
```
Comment (oshadmi) [inline src/spec.c:1760]: Review by Meir: need to cleanup all `IndexSpec` created in this call when a short-read occurs,
e.g., if 2 `IndexSpec` were read successfully and the third one fails due to short-read, need to cleanup all three.

Comment (tw-bert) [reply src/spec.c:1760]: @oshadmi Is it possible to give an ETA (roughly)?

Comment (oshadmi) [reply src/spec.c:1766]: @tw-bert I apologize for my late response. ETA hopefully this week (or maybe next week) once code is reviewed and probably amended accordingly.

Comment (tw-bert) [reply src/spec.c:1766]: @oshadmi Thanks for the update!

File: .circleci/ci_coverage.sh [hunk 2]
```diff
@@ -23,6 +23,9 @@ cat >rltest.config <<EOF
 --unix
 EOF
 export CONFIG_FILE="$PWD/rltest.config"
+export DIAG_CODE_COVERAGE=1
+export VERBOSE=1
+export REJSON_BRANCH=omer-short-read
```
Comment (rafie) [inline .circleci/ci_coverage.sh:28]: we need to merge the RedisJSON short-reads PR and to clean up here.

File: .circleci/ci_diag_san12.sh [hunk 3]
```diff
@@ -71,5 +71,6 @@ source /etc/profile.d/rust.sh
 make nightly
 make SAN=$JSON_SAN_MODE
 export REJSON_PATH=$ROOT/deps/RedisJSON/target/x86_64-unknown-linux-gnu/debug/rejson.so
+export DIAG_SANITIZER=1
```
Comment (rafie) [inline .circleci/ci_diag_san12.sh:74]: remove?

File: .circleci/config.yml [hunk 4]
```diff
@@ -199,12 +201,13 @@ jobs:
       - run:
           name: Run & build tests
           command: bash -l -c "MSAN=1 ./.circleci/ci_diag_san12.sh"
-          no_output_timeout: 20m
+          timeout: 180m
+          no_output_timeout: 180m
```
Comment (rafie) [inline .circleci/config.yml:205]: We can keep the original timeouts.

File: .circleci/config.yml [hunk 5]
```diff
@@ -187,7 +188,8 @@ jobs:
       - run:
           name: Run & build tests
           command: bash -l -c "ASAN=1 ./.circleci/ci_diag_san12.sh"
-          no_output_timeout: 20m
+          timeout: 180m
+          no_output_timeout: 180m
```
Comment (rafie) [inline .circleci/config.yml:192]: We can keep the original timeouts.

File: src/notifications.c [hunk 6]
```diff
@@ -311,3 +313,55 @@ void Initialize_CommandFilter(RedisModuleCtx *ctx) {
...
+    Discard_Globals();
+    break;
+  }
+}
+
+void LoadingCallback(RedisModuleCtx *ctx, RedisModuleEvent eid, uint64_t subevent, void *data) {
+
+  REDISMODULE_NOT_USED(eid);
+  if (subevent == REDISMODULE_SUBEVENT_LOADING_FAILED) {
+      Discard_Global_IndexSpecs();
```
Comment (oshadmi) [inline src/notifications.c:338]: Why not also `Discard_Globals()` ?

**Review by oshadmi — COMMENTED** [review]
Comments from code review meeting - will continue later

**Review by oshadmi — COMMENTED** [review]
Few more comments from a code review session today requiring some changes.

### Additional discussion

- Investigate search failures after a full read:

```
# ExpectedIndex(2, 'shortread_idxJson_[1-9]', [0, 0]),  #FIXME: count should be [20,55]
# ExpectedIndex(2, 'shortread_idxSearchJson_[1-9]', [10, 0])  #FIXME: count should be [20,55]

```   

- Review valgrind report from test scenario

(Ext_RegisterScoringFunction, Ext_RegisterQueryExpander, ...)

How not found by MSAN Sanitizer?
Can compare to valgrind on master (with a simple scenario of module load)


### Once merged

- Add test for sanitizer

- Move python test code to common place (for ReJSON or other modules)

Review the impact (and risk) of using existing macros such as `array_new_on_stack` in [Gears](https://github.com/RedisGears/RedisGears/blob/master/src/utils/arr.h) - instead of specific code in [rules.c](https://github.com/RediSearch/RediSearch/pull/2058/files#diff-2b582c78ca608ab7598a23b59a0e648207c56589f0f40eae8818d3a6643a8a41) in `SchemaRule_RdbLoad` (using `RULEARGS_INITIAL_NUM_PREFIXES_ON_STACK`)

File: src/spec.c [hunk 7]
```diff
@@ -1126,25 +1127,29 @@ int bit(t_fieldMask id) {
 ///////////////////////////////////////////////////////////////////////////////////////////////
 
 // Backwards compat version of load for rdbs with version < 8
-static void FieldSpec_RdbLoadCompat8(RedisModuleIO *rdb, FieldSpec *f, int encver) {
-  RedisModule_LoadStringBufferAlloc(rdb, f->name, NULL);
+static int FieldSpec_RdbLoadCompat8(RedisModuleIO *rdb, FieldSpec *f, int encver) {
+  LoadStringBufferAlloc_IOErrors(rdb, f->name, NULL, goto fail);
```
[the remark this answers is not in the document]
Comment (oshadmi) [reply src/spec.c:1129]: @rafie in `LoadStringBufferAlloc_IOErrors`, if `oldbuf` is `NULL` then IO Error occurred and `cleanup_exp` is executed (which should goto and skip the rest of the macro) and will not get to the `memcpy`.
We can add an assert `assert(oldbuf);`

```
...
char *oldbuf = RedisModule_LoadStringBuffer(rdb, tmp_len_ptr); \
if (RedisModule_IsIOError(rdb)) { \
    cleanup_exp; \
} \
assert(oldbuf); \
...
```

File: src/spec.c [hunk 8]
```diff
@@ -1560,6 +1570,8 @@ IndexSpec *IndexSpec_CreateFromRdb(RedisModuleCtx *ctx, RedisModuleIO *rdb, int
 
   if (sp->flags & Index_HasCustomStopwords) {
     sp->stopwords = StopWordList_RdbLoad(rdb, encver);
+    if (sp->stopwords == NULL)
```
Comment (MeirShpilraien) [inline src/spec.c:1573]: Are we sure NULL is not a valid stopwords? We did not have this check before.

File: src/spec.c [hunk 9]
```diff
@@ -1571,29 +1583,29 @@ IndexSpec *IndexSpec_CreateFromRdb(RedisModuleCtx *ctx, RedisModuleIO *rdb, int
   CursorList_AddSpec(&RSCursors, sp->name, RSCURSORS_DEFAULT_CAPACITY);
   RedisModule_FreeString(ctx, specKey);
 
-  sp->smap = NULL;
   if (sp->flags & Index_HasSmap) {
     sp->smap = SynonymMap_RdbLoad(rdb, encver);
+    if (sp->smap == NULL)
```
Comment (MeirShpilraien) [inline src/spec.c:1588]: Same question as with stopwords

File: src/spec.c [hunk 10]
```diff
@@ -1571,29 +1583,29 @@ IndexSpec *IndexSpec_CreateFromRdb(RedisModuleCtx *ctx, RedisModuleIO *rdb, int
...
-  size_t narr = RedisModule_LoadUnsigned(rdb);
+  size_t narr = LoadUnsigned_IOError(rdb, goto cleanup);
   for (size_t ii = 0; ii < narr; ++ii) {
     QueryError _status;
     size_t dummy;
-    char *s = RedisModule_LoadStringBuffer(rdb, &dummy);
+    char *s = LoadStringBuffer_IOError(rdb, &dummy, goto cleanup);
     int rc = IndexAlias_Add(s, sp, 0, &_status);
     RedisModule_Free(s);
     RedisModule_Log(NULL, "notice", "Loading existing alias failed");
```
Comment (MeirShpilraien) [inline src/spec.c:1604]: Why do we always print this log message?

File: src/spec.c [hunk 11]
```diff
@@ -1612,6 +1624,12 @@ IndexSpec *IndexSpec_CreateFromRdb(RedisModuleCtx *ctx, RedisModuleIO *rdb, int
   }
 
   return sp;
+
+cleanup:
+  IndexSpec_Free(sp);
+  Indexes_Free(specDict_g);
```
Comment (MeirShpilraien) [inline src/spec.c:1630]: Isn't `specDict_g` is the global spec dictionary? Why do you free it here?

File: src/spec.c (hunk 8 above)
Comment (MeirShpilraien) [reply src/spec.c:1573]: Got it, the old function never returned NULL :+1:

File: src/spec.c (hunk 9 above)
Comment (MeirShpilraien) [reply src/spec.c:1588]: Same as above :+1:

File: tests/pytests/test_short_read.py [hunk 12]
```diff
@@ -0,0 +1,566 @@
...
+            self.dbg_str = self.dbg_str + "\n"
+        self.dbg_ndx = self.dbg_ndx + 1
+
+        env.debugPrint(name + ': %d out of %d \n%s' % (self.dbg_ndx, total_len, self.dbg_str))
+
+
+def testShortReadSearch(env):
+    if IS_SANITIZER:
+        env.skip()  # Sanitizer is taking too long - should be done on a separate/dedicated test job
+
```
Comment (MeirShpilraien) [inline tests/pytests/test_short_read.py:455]: Also skip on cluster and on existing env

File: src/spec.c (hunk 10 above)
Comment (oshadmi) [reply src/spec.c:1604]: Seems to be done unintentionally by PR #2045

@ashtul We should add a condition to the log, agree?
```
if (rc != REDISMODULE_OK) {
  RedisModule_Log(NULL, "notice", "Loading existing alias failed");
}

```

**Comment by oshadmi** [comment]
Additional comments during review with Meir:

- Make Sanitizer work - cannot merge without it

- Skip test on cluster and on existing-env

File: src/notifications.c [hunk 13]
```diff
@@ -334,12 +334,11 @@ void ReplicaBackupCallback(RedisModuleCtx *ctx, RedisModuleEvent eid, uint64_t s
 
 
 int CheckVersionForShortRead() {
-  // Minial versions: 6.2.5 or 6.0.15
-  if (redisVersion.majorVersion == 6) {
-    if (redisVersion.minorVersion == 2)
+  // Minial versions: 6.2.5
```
Comment (tw-bert) [inline src/notifications.c:337]: -minor typo- (Minimal) Thx for the clear commit comment, now I know we need 6.2.5, soonish. We're already at 6.2.4, so no issues expected on our side. Cheers TW

**Review by rafie — APPROVED** [review]

**Pull request merged**